sdk/core/azure-core/inc/azure/core/internal/client_options.hpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Base type for all client option types, exposes various common client options like Retry
+ * and Transport.
+ */
+
+#pragma once
+
+#include "azure/core/http/http.hpp"
+#include "azure/core/http/policy.hpp"
+
+#include <memory>
+#include <vector>
+
+namespace Azure { namespace Core { namespace Internal {
+
+  /**
+   * @brief Represents a position of the policy in the pipeline.
+   *
+   */
+  enum class HttpPipelinePosition
+  {
+    /**
+     * @brief The policy would be invoked once per pipeline invocation (service call).
+     *
+     */
+    PerCall,
+    /**
+     * @brief The policy would be invoked every time request is retried.
+     *
+     */
+    PerRetry
+  };
+
+  /**
+   * @brief  Base type for all client option types, exposes various common client options like Retry
+   * and Transport.
+   *
+   */
+  class ClientOptions {
```

Comment:
just for the record, I still think

```
struct ClientOptions {
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perOperationPolicies;
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perRetryPolicies;
    Azure::Core::Http::TransportPolicyOptions Transport;
    ...
};
```

would be better.

Other client options are gonna inherit from this class. A struct with member functions looks kind of weird

Reply:
But this is not struct. This is a class which has an encapsulated way to AddPolicies()

Reply:
yeah, so `KeyClientOptions` is also going to be a `class` maybe? I dont know.

I also wrote some struct with member functions. Then someone (I cannot remember who) jumped out and said I cannot do that.

Reply:
I think you would do the same for any storage client option which you would like to keep apart from its implementation details.

Reply:
I don't like that and I don't think that'd be necessary and I don't see that'd benefit our customers. But I will make the change to be consistent with core sdk.